kernel/vmbytegen.ml

```diff
@@ -629,38 +631,35 @@ let rec compile_lam env cenv lam sz cont =
       let lbl_bodies = Array.make ndef Label.no in
       (* Compiling types *)
       let rfv = ref empty_fv in
-      let env_type = comp_env_cofix_type ndef rfv in
       for i = 0 to ndef - 1 do
-        let fcode =
-          ensure_stack_capacity (compile_lam env env_type types.(i) 0) [Kstop]
-        in
+        let env_type = comp_env_cofix_type ndef rfv in
+        let fcode = compile_lam env env_type types.(i) 0 [Kstop] in
+        let fcode = ensure_stack_capacity env_type fcode in
         let lbl,fcode = label_code fcode in
         lbl_types.(i) <- lbl;
-        fun_code := Ksequence fcode :: !fun_code
+        push_fun env fcode
       done;
       (* Compiling bodies *)
       for i = 0 to ndef - 1 do
         let params,body = decompose_Llam bodies.(i) in
         let arity = Array.length params in
         let env_body = comp_env_cofix ndef arity rfv in
         let lbl = Label.create () in
-        let comp arity =
-          (* 4 stack slots are needed to update the cofix when forced *)
-          set_max_stack_size (arity + 4);
-          compile_lam env env_body body (arity+1) (cont_cofix arity)
-        in
-        let cont = ensure_stack_capacity comp arity in
+        (* 4 stack slots are needed to update the cofix when forced *)
+        let () = set_max_stack_size env_body (arity + 4) in
+        let cont = compile_lam env env_body body (arity+1) (cont_cofix arity) in
+        let cont = ensure_stack_capacity env_body cont in
         lbl_bodies.(i) <- lbl;
-        fun_code := Ksequence (add_grab (arity+1) lbl cont) :: !fun_code;
+        push_fun env (add_grab (arity+1) lbl cont)
       done;
       let fv = !rfv in
-      set_max_stack_size (sz + fv.size + ndef + 2);
+      let () = set_max_stack_size cenv (sz + fv.size + ndef + 2) in
```

Comment:
Isn't this still a sequence? Why change it to a let binding?

Reply:
I personally hate the sequence operator in OCaml, due to parsing weirdness it's way to easy to break with innocuous code manipulations. Furthermore, this way of writing things makes it very explicit that we're doing this for the side-effect.